go/processor/floria/precompiled.go

```diff
@@ -48,3 +54,215 @@ func getPrecompiledContract(address tosca.Address, revision tosca.Revision) (get
+
+var (
+	setBalanceMethodID []byte
+	copyCodeMethodID   []byte
+	swapCodeMethodID   []byte
+	setStorageMethodID []byte
+	incNonceMethodID   []byte
+)
+
+func init() {
+	abi, err := abi.JSON(strings.NewReader(stateContractABI))
+	if err != nil {
+		panic(err)
+	}
+
+	for name, constID := range map[string]*[]byte{
+		"setBalance": &setBalanceMethodID,
+		"copyCode":   &copyCodeMethodID,
+		"swapCode":   &swapCodeMethodID,
+		"setStorage": &setStorageMethodID,
+		"incNonce":   &incNonceMethodID,
+	} {
+		method, exist := abi.Methods[name]
+		if !exist {
+			panic("unknown EvmWriter method")
+		}
+
+		*constID = make([]byte, len(method.ID))
+		copy(*constID, method.ID)
+	}
+}
+
+var ErrExecutionReverted = fmt.Errorf("execution reverted")
+var ErrOutOfGas = fmt.Errorf("out of gas")
+
+func handleStatePrecompiledContract(state tosca.WorldState, sender tosca.Address, address tosca.Address, input []byte, gas tosca.Gas) (tosca.CallResult, bool) {
+	if address != StateContractAddress {
+		return tosca.CallResult{}, false
+	}
+	gas, err := statePrecompiledContract(state, sender, input, gas)
+	return tosca.CallResult{
+		Success: err == nil,
+		Output:  nil,
+		GasLeft: tosca.Gas(gas),
+	}, true
+}
+
+func statePrecompiledContract(state tosca.WorldState, sender tosca.Address, input []byte, gas tosca.Gas) (tosca.Gas, error) {
+	if sender != DriverAddress {
+		return 0, ErrExecutionReverted
+	}
+	if len(input) < 4 {
+		return 0, ErrExecutionReverted
+	}
+	selector := input[:4]
+	input = input[4:]
+	if bytes.Equal(selector, setBalanceMethodID) {
+		return stateSetBalance(state, sender, input, gas)
+	} else if bytes.Equal(selector, copyCodeMethodID) {
+		return stateCopyCode(state, input, gas)
+	} else if bytes.Equal(selector, swapCodeMethodID) {
+		return stateSwapCode(state, input, gas)
+	} else if bytes.Equal(selector, setStorageMethodID) {
+		return stateSetStorage(state, input, gas)
+	} else if bytes.Equal(selector, incNonceMethodID) {
+		return stateIncNonce(state, sender, input, gas)
+	}
+
+	return 0, fmt.Errorf("invalid method ID")
+}
+
+func stateSetBalance(state tosca.WorldState, sender tosca.Address, input []byte, gas tosca.Gas) (tosca.Gas, error) {
+	if gas < CallValueTransferGas {
+		return 0, ErrOutOfGas
+	}
+	gas -= CallValueTransferGas
+	if len(input) != 64 {
+		return 0, ErrExecutionReverted
+	}
+
+	account := tosca.Address(input[12:32])
+	value := tosca.Value(input[32:64])
+
+	if account == sender {
+		// Origin balance shouldn't decrease during his transaction
+		return 0, ErrExecutionReverted
+	}
+
+	state.SetBalance(account, value)
+	return gas, nil
+}
+
+func stateCopyCode(state tosca.WorldState, input []byte, gas tosca.Gas) (tosca.Gas, error) {
+	if gas < CreateGas {
+		return 0, ErrOutOfGas
+	}
+	gas -= CreateGas
+	if len(input) != 64 {
+		return 0, ErrExecutionReverted
+	}
+
+	accountTo := tosca.Address(input[12:32])
+	accountFrom := tosca.Address(input[32+12 : 32+32])
+
+	code := state.GetCode(accountFrom)
+	if code == nil {
```

Comment:
curious: is this check needed or would the rest of the code also work with a `nil` value for the code?

Reply:
I removed the checks and added an unit test to ensure it works as expected.